cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -74,88 +83,160 @@
+    typesToKeys.put(ARTIFACT, new String[][]{{NAMESPACE, ARTIFACT, VERSION}});
+    typesToKeys.put(VIEW, new String[][]{{NAMESPACE, STREAM, VIEW}});
+    typesToKeys.put(PROGRAM, new String[][]{{NAMESPACE, APPLICATION, VERSION, TYPE, PROGRAM},
+      {NAMESPACE, APPLICATION, TYPE, PROGRAM}});
+    typesToKeys.put(SCHEDULE, new String[][]{{NAMESPACE, APPLICATION, VERSION, SCHEDULE},
+      {NAMESPACE, APPLICATION, SCHEDULE}});
+    typesToKeys.put(FLOWLET, new String[][]{{NAMESPACE, APPLICATION, VERSION, FLOW, FLOWLET},
+      {NAMESPACE, APPLICATION, FLOW, FLOWLET}});
+    typesToKeys.put(PROGRAM_RUN, new String[][]{{NAMESPACE, APPLICATION, VERSION, TYPE, PROGRAM, PROGRAM_RUN},
+      {NAMESPACE, APPLICATION, TYPE, PROGRAM, PROGRAM_RUN}});
+    TYPES_TO_KEY_SEQUENCES = Collections.unmodifiableMap(typesToKeys);
   }
 
   /**
-   * Creates a {@link MetadataEntity} representing the given datasetName. To create a {@link MetadataEntity} for a
-   * dataset in a specified namespace please use {@link MetadataEntity#ofDataset(String, String)}.
-   *
-   * @param datasetName the name of the dataset
-   * @return {@link MetadataEntity} representing the dataset name
+   * Used by Builder to build the {@link MetadataEntity}
    */
-  public static MetadataEntity ofDataset(String datasetName) {
-    return new MetadataEntity(DATASET, datasetName);
+  private MetadataEntity(Map<String, String> details, String type) {
+    this.details = details;
+    this.type = type;
   }
 
   /**
-   * Creates a {@link MetadataEntity} representing the given datasetName in the specified namespace.
-   *
-   * @param namespace the name of the namespace
-   * @param datasetName the name of the dataset
-   * @return {@link MetadataEntity} representing the dataset name
+   * Builder for {@link MetadataEntity}
    */
-  public static MetadataEntity ofDataset(String namespace, String datasetName) {
-    return MetadataEntity.ofNamespace(namespace).appendAsType(MetadataEntity.DATASET, datasetName);
+  public static class Builder {
+    private final LinkedHashMap<String, String> parts;
+    private String type;
+
+    Builder() {
+      parts = new LinkedHashMap<>();
+    }
+
+    private Builder(MetadataEntity metadataEntity) {
+      this.parts = new LinkedHashMap<>(metadataEntity.details);
+      this.type = metadataEntity.type;
+    }
+
+    /**
+     * Put the given key with the given value in {@link MetadataEntity.Builder}.
+     * The returned {@link MetadataEntity.Builder} is of the same type. If the type needs to
+     * changed during append then {@link #appendAsType(String, String)}  should be used.
+     *
+     * @param key the key to be added
+     * @param value the value to be added
+     * @return a new {@link MetadataEntity.Builder} which is of same type of this {@link MetadataEntity.Builder}
+     * but consists of the given key and value in addition
+     */
+    public Builder append(String key, String value) {
+      parts.put(key.toLowerCase(), value);
```

Comment:
I think the Javadocs should mention that the keys are not case-sensitive.

Reply:
Done.